Add REPL documentation to the Gatsby Docs

n8finch opened the pull request:
## Description

Currently [WIP], this PR creates documentation for the Gatsby CLI REPL shell. This was merged in #7262, but didn't include documentation.

I've labeled this [WIP] because I'd like feedback on the format and content of the documentation:

1. Are the explanations clear?
2. Are the examples helpful?

I just want to make sure 🤓.

Once I get feedback, I'll finish up the docs for a merge.

Also, it looks like the `dataPaths` command is throwing an error? But I'll double check on that and file a bug issue (and hopefully a PR to fix it!) after the docs.

## Related Issues

Fixes #7484

lannonbr commented:
Thanks for the work so far @n8finch! Just a thing to do is since you are fleshing out this stub, I would recommend removing the asterisk for the `gatsby-repl` entry in the sidebar as it is no longer going to be a stub. It can be found in the sidebar yaml file located in `www/src/data/sidebars/doc-links.yaml`. Here's the file in the master branch: https://github.com/gatsbyjs/gatsby/blob/master/www/src/data/sidebars/doc-links.yaml

n8finch commented:
@lannonbr thanks for the heads up, I'd completely missed that! 👍

n8finch commented:
Hmm, I'm showing some CircleCI tests failing... When I look at the details, I'm not sure what went wrong. Do I need to run `gatsby build` to compile in the yaml updates?

lannonbr commented:
> Hmm, I'm showing some CircleCI tests failing... When I look at the details, I'm not sure what went wrong. Do I need to run `gatsby build` to compile in the yaml updates?

So there's 502 bad gateway errors coming up. Cloudflare is having some outages today and NPM is affected by such, so it's no issue on your end

gillkyle commented on docs/docs/gatsby-repl.md:
```diff
@@ -1,9 +1,133 @@
 ---
 
-This is a stub. Help our community expand it.
+The Gatsby REPL ("read-eval-print loop") is available via the Gatsby CLI by running the command `gatsby repl`. This gives you access to an interactive REPL shell within the context of your Gatsby environment. It can be used to retrived general data and programatically interact with it. If you have an idea for a new command, feel free to submit a PR for it!
```
```suggestion
The Gatsby REPL ("read-eval-print loop") is available via the Gatsby CLI by running the command `gatsby repl`. This gives you access to an interactive REPL shell within the context of your Gatsby environment. It can be used to retrieve general data and programmatically interact with it. If you have an idea for a new command, feel free to submit a PR for it!
```

gillkyle commented on packages/gatsby-cli/src/create-cli.js:
```diff
@@ -266,7 +266,7 @@ function buildLocalCommands(cli, isLocalSite) {
   cli.command({
     command: `repl`,
-    desc: `Get a node repl with context of Gatsby environment, see (add docs link here)`,
+    desc: `Get a node repl with context of Gatsby environment, see (https://www.gatsbyjs.org/docs/gatsby-repl/)`,
```
```suggestion
    desc: `Get a node repl with context of Gatsby environment, see (/docs/gatsby-repl/)`,
```

This can just be a relative link 🙂

n8finch replied:
Hey @gillkyle, thanks for the review and the corrections! I've never been the best speller 😂.

I'm pushing up a new commit now with those changes. I'll get the WIP removed and add more documentation here over the next couple of days.

Thanks again!

n8finch commented:
Hey @gillkyle, thanks for the review and the corrections! I've never been the best speller 😂.

I've pushed up a new commit now with those changes. 

I've also removed the WIP and added more documentation here.

It looks like there might be some tests failing, which I'm not sure how to resolve. Any pointers?

Thanks again!

gillkyle commented:
Awesome! Thanks @n8finch! Those additional updates look awesome 😄 

As far as tests go, it looks like CircleCI is complaining about linting, I think the command from [this doc](https://www.gatsbyjs.org/contributing/managing-pull-requests/#linting) might be enough:

```
npm run format
```

and then push the changes. Seems like the offending file is `docs/docs/gatsby-repl.md`.

I've seen the windows tests just kind of fail for no real obvious reason, so pushing some change might just be enough for them to run again and maybe work. Might have to look into those more if they keep failing.

n8finch commented:
@gillkyle Woohoo! That worked... Hahaha.. Machines... Thanks!

gillkyle commented on packages/gatsby-cli/src/create-cli.js:
```diff
@@ -266,7 +266,7 @@ function buildLocalCommands(cli, isLocalSite) {
   cli.command({
     command: `repl`,
-    desc: `Get a node repl with context of Gatsby environment, see (add docs link here)`,
+    desc: `Get a node repl with context of Gatsby environment, see (/docs/gatsby-repl/)`,
```
```suggestion
    desc: `Get a node repl with context of Gatsby environment, see (https://www.gatsbyjs.org/docs/gatsby-repl/)`,
```

Because this one is output from the CLI it can be the full path 🙂

gillkyle replied:
Just going to commit this myself though 👍 thanks for all the work on this!

n8finch replied:
Ahh, gotcha, thanks for pointing that out! Add that to my most recent commit.

muescha commented on docs/docs/gatsby-repl.md:
```diff
@@ -1,9 +1,215 @@
+
+### `components`
+
+Returns a Map object with all of the components in your Gatsby environment (see [Mozilla Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) docs for more information on Map objects and how to use them). Propeties that get returned: `name`, `componentPath`, `id`, `query`, and `hash`.
```
also property `pages` (used in the code snippet below)?

n8finch replied:
Thanks for pointing that out! I changed this in the most recent commit.

n8finch commented:
Thanks for your help @gillkyle , @lannonbr , and @muescha ! I learned a lot!
